### PR TITLE
Chapter12: Reset password user using mail...

### PR DESCRIPTION
**Pull request by giapbnv-0886**

### CODE REVIEW

**Comment by giapbnv-0886** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
+  end
+
+  def valid_user
+    unless (@user && @user.activated? && @user.authenticated?(:reset, params[:id]))
```
Comment (datdt-0881) [inline app/controllers/password_resets_controller.rb]: &.

File: app/models/user.rb [hunk 2]
```diff
@@ -40,7 +40,21 @@ def activate
 
   def send_activation_email
     UserMailer.account_activation(self).deliver_now
-    end
+  end
+
+  def create_reset_digest
+    self.reset_token = User.new_token
+    update_attribute :reset_digest, User.digest(reset_token)
+    update_attribute :reset_sent_at, Time.zone.now
```
Comment (datdt-0881) [inline app/models/user.rb]: update_attributes

File: app/views/password_resets/edit.html.erb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<% provide :title, t("users.mail.resetpassword")  %>
+<h1><%= t "users.mail.resetpassword" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(@user, url: password_reset_path(params[:id])) do |f| %>
```
Comment (datdt-0881) [inline app/views/password_resets/edit.html.erb]: ()

File: app/views/password_resets/new.html.erb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+<% provide :title, t("users.forgot.title") %>
+<h1><%=t "users.forgot.title" %></h1>
+
+<div class="row">
+  <div class="col-md-6 col-md-offset-3">
+    <%= form_for(:password_reset, url: password_resets_path) do |f| %>
```
Comment (datdt-0881) [inline app/views/password_resets/new.html.erb]: ()

**Comment by giapbnv-0886** [comment]
ready

File: app/controllers/password_resets_controller.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
+class PasswordResetsController < ApplicationController
+  before_action :get_user, only: %i(edit update)
+  before_action :valid_user, only: %i(edit update)
+  before_action :check_expiration, only: %i(edit, update)
+
+  def new
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]:
```suggestion
  def new; end
```

File: app/controllers/password_resets_controller.rb [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  private
+
+  def user_params
+    params.require(:user).permit :password, :password_confirmation
+  end
+
+  def get_user
+    @user = User.find_by email: params[:email]
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: check đk k tìm thấy @user

File: app/controllers/password_resets_controller.rb [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  end
+
+  def valid_user
+    unless @user && @user.activated? && @user.authenticated?(:reset, params[:id])
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: return if

File: app/controllers/password_resets_controller.rb [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+      flash[:danger] = t "users.forgot.token_invalid"
+      redirect_to root_path
+    end
+  end
+
+  def check_expiration
+    if @user.password_reset_expired?
```
Comment (huuphuc93) [inline app/controllers/password_resets_controller.rb]: return unless

File: app/models/user.rb [hunk 9]
```diff
@@ -30,17 +30,30 @@ def authenticated? attribute, token
...
+    update_attributes reset_digest: User.digest(reset_token),
+      reset_sent_at: Time.zone.now
+  end
+
+  def send_password_reset_email
+    UserMailer.password_reset(self).deliver_now
+  end
+
+  def password_reset_expired?
+    reset_sent_at < 2.hours.ago
```
Comment (huuphuc93) [inline app/models/user.rb]: 2 => Settings

**Comment by giapbnv-0886** [comment]
ready

**Comment by huuphuc93** [comment]
merged

**Pull request merged**